### PR TITLE
Feature/2729 survey report

### PR DESCRIPTION
**Pull request by borkarsaish65**

### CODE REVIEW

File: common/files_helper.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+module.exports = {
```
Comment (VISHNUDAS-tunerlabs) [inline common/files_helper.js:2]: @borkarsaish65 why we need separate file for this? generic->messageconstants we already have . Please re use them. IF anything missing add there

File: controllers/v1/reportsController.js [hunk 2]
```diff
@@ -3012,4 +3017,238 @@ module.exports = class Reports {
...
+ *     }
+ *   ],
+ *   "status": 200
+ * }
+ */
+
+submissionReport = async function (req, res) {
+  return new Promise(async function (resolve, reject) {
+    try {
+      if (!req.query.submissionId) {
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reportsController.js:3065]: @borkarsaish65 Please do not have process logic inside controller. move them to module helpers

File: controllers/v1/reportsController.js [hunk 3]
```diff
@@ -3012,4 +3017,238 @@ module.exports = class Reports {
...
+        message: err.message || 'INTERNAL_SERVER_ERROR',
+      };
+      return resolve(response);
+    }
+  });
+};
+
 };
+
+function transformData(data) {
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reportsController.js:3157]: @borkarsaish65 function signature and description is missing. Please add

File: controllers/v1/reportsController.js [hunk 4]
```diff
@@ -3012,4 +3017,238 @@ module.exports = class Reports {
...
+        answerOptions: singleAnswerObj.options ? singleAnswerObj.options : undefined,
+      },
+    });
+  });
+
+  return singleQuestionAnswerDataPerSolutionArr;
+  //return data;
+}
+
+function formateEvidenceData(surveyAnswers, chartData) {
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reportsController.js:3217]: @borkarsaish65 why we are adding these functions in controller? Are these new endpoints? Controller should call module helper and get the data and return the response. All other logics we should not have in controllers

File: generics/helpers/utils.js [hunk 5]
```diff
@@ -204,6 +204,11 @@ function filterLocationIdandCode(dataArray) {
 function arrayIdsTobjectIds(ids) {
   return ids.map((id) => ObjectId(id));
 }
+function arrayIdsTobjectIdsNew(ids) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:207]: @borkarsaish65 please add function descriptions

File: helper/chart_data.js [hunk 6]
```diff
@@ -0,0 +1,2788 @@
+
+const path = require("path");
+
+const default_no_of_assessment_submissions_threshold = 3;
+const default_entity_score_api_threshold = 5;
+const filesHelper = require(MODULES_BASE_PATH + '/cloud-services/files/helper')
+
+//function for instance observation final response creation
+exports.instanceReportChart = async function (data, reportType = "") {
```
Comment (VISHNUDAS-tunerlabs) [inline helper/chart_data.js:9]: @borkarsaish65 Please add function description where ever it is missing

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Reviewed on Aug 19 5:30

File: controllers/v1/reportsController.js (hunk 3 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js:3082]: I have added now @VISHNUDAS-tunerlabs

File: controllers/v1/reportsController.js (hunk 4 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js:3142]: changed now @VISHNUDAS-tunerlabs

File: generics/helpers/utils.js (hunk 5 above)
Comment (borkarsaish65) [reply generics/helpers/utils.js:210]: added

File: common/files_helper.js (hunk 1 above)
Comment (borkarsaish65) [reply common/files_helper.js]: done

File: controllers/v1/reportsController.js (hunk 2 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js]: done

File: controllers/v1/reportsController.js (hunk 3 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js]: rewrote entire logic

File: controllers/v1/reportsController.js (hunk 4 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js]: not used anymore so didnt make changes

File: helper/chart_data.js (hunk 6 above)
Comment (borkarsaish65) [reply helper/chart_data.js:20]: done

File: controllers/v1/reportsController.js [hunk 7]
```diff
@@ -11,6 +11,11 @@ const FileStream = require(ROOT_PATH + '/generics/fileStream');
 const solutionsHelper = require(MODULES_BASE_PATH + '/solutions/helper');
 const reportsHelper = require(MODULES_BASE_PATH + '/reports/helper');
 let imageBaseUrl = '';
+const surveySubmissionsHelper = require(MODULES_BASE_PATH + '/surveySubmissions/helper');
+const questionsHelper = require(MODULES_BASE_PATH + '/questions/helper');
+const programsHelper = require(MODULES_BASE_PATH + '/programs/helper');
+const helperFunc = require('../../helper/chart_data');
+const solutionsQueries = require(DB_QUERY_BASE_PATH + '/solutions');
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reportsController.js:18]: @borkarsaish65 we need this? are we using it in controller? we should not have db query called in controllers

File: controllers/v1/reportsController.js [hunk 8]
```diff
@@ -3012,4 +3017,64 @@ module.exports = class Reports {
...
+
+submissionReport = async function (req) {
+  return new Promise(async function (resolve, reject) {
+    try {
+
+      let generateSubmissionReport = await reportsHelper.surveySubmissionReport(req);
+      resolve(generateSubmissionReport);
+
+    } catch (err) {
+      console.log(err)
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reportsController.js:3070]: @borkarsaish65 remove console if not required

File: generics/helpers/utils.js [hunk 9]
```diff
@@ -202,8 +202,16 @@ function filterLocationIdandCode(dataArray) {
 }
 
 function arrayIdsTobjectIds(ids) {
+  /**
+ * Converts an array of IDs into an array of `ObjectId` instances.
+ */
   return ids.map((id) => ObjectId(id));
 }
+function arrayIdsTobjectIdsNew(ids) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:210]: @borkarsaish65 Please add function desription

File: helper/chart_data.js [hunk 10]
```diff
@@ -0,0 +1,2932 @@
...
+            // Get the questions array
+            let questionArray = await questionListObjectCreation(actualData);
+            result.allQuestions = questionArray;
+        }
+
+        //return final response object
+        return result;
+    }
+    catch (err) {
+        console.log(err);
```
Comment (VISHNUDAS-tunerlabs) [inline helper/chart_data.js:160]: @borkarsaish65 Are we just consoling the error?

File: helper/chart_data.js [hunk 11]
```diff
@@ -0,0 +1,2932 @@
...
+    }
+}
+
+
+//Function to create a response object for multiselect question (Instance level Report)
+async function instanceMultiselectFunc(data) {
```
Comment (VISHNUDAS-tunerlabs) [inline helper/chart_data.js:166]: @borkarsaish65 Please add proper function description ( Please check for all other functions in this helper)

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on 27 Aug : 12

File: generics/helpers/utils.js (hunk 9 above)
Comment (borkarsaish65) [reply generics/helpers/utils.js:220]: done

File: controllers/v1/reportsController.js (hunk 8 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js]: done

File: helper/chart_data.js (hunk 10 above)
Comment (borkarsaish65) [reply helper/chart_data.js]: done

File: helper/chart_data.js (hunk 11 above)
Comment (borkarsaish65) [reply helper/chart_data.js]: this is removed as it was not used yet and was moved frm ml-survey

File: controllers/v1/reportsController.js (hunk 7 above)
Comment (borkarsaish65) [reply controllers/v1/reportsController.js]: removed now

**Pull request merged**